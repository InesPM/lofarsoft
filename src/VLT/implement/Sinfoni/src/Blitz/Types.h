//#  Types.h : Make Blitz-specific typedefs available in global namespace.
//#
//#  Copyright (C) 2002-2006
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: Types.h,v 1.5 2006/05/19 16:03:23 loose Exp $

#ifndef SINFONI_BLITZ_TYPES_H
#define SINFONI_BLITZ_TYPES_H

//# Includes
#include <Sinfoni/Blitz/Typedefs.h>
#include <Sinfoni/Types.h>

namespace ASTRON
{
  namespace SINFONI
  {
    // @name Types
    // Make the type names defined in Blitz/Typedefs.h available in the 
    // namespace ASTRON::SINFONI.
    //@{
    using TYPES::VectorInteger;
    using TYPES::VectorReal;
    using TYPES::VectorDouble;
    using TYPES::VectorComplex;
    using TYPES::VectorDComplex;
    
    using TYPES::MatrixInteger;
    using TYPES::MatrixReal;
    using TYPES::MatrixDouble;
    using TYPES::MatrixComplex;
    using TYPES::MatrixDComplex;
    //@}

  } // namespace SINFONI

} // namespace ASTRON

#endif
