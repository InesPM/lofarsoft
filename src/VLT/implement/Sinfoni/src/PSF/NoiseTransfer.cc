//#  NoiseTransfer.cc: global methods needed to calculate the noise transfer.
//#
//#  Copyright (C) 2002-2006
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: NoiseTransfer.cc,v 1.10 2006/05/19 16:03:24 loose Exp $

#include <Sinfoni/PSF/NoiseTransfer.h>
#include <Sinfoni/Math.h>
#include <Sinfoni/Assert.h>
#include <Sinfoni/Debug.h>

using namespace std;

namespace ASTRON
{
  namespace SINFONI
  {
    namespace PSF
    {
      // This unnamed namespace bundles the methods required to compute the
      // noise transfer.
      namespace
      {
        // The integration time \f$T_e\f$.
        const Real integrationTime = 1.0/350;

        // The loop delay \f$\tau\f$.
        const Real loopDelay = 0.31e-3;

        // Compute the system transfer function, which is defined as:
        // \f[
        //   H_{sys}(f,g_i) = \frac{g_i}{2\pi i f T_e} \exp{(-2\pi i f \tau)}
        // \f]
        // where \f$f\f$ is the frequency \a freq, \f$g_i\f$ is the gain of
        // the electrode \a gain, \f$T_e\f$ is the integration time, and
        // \f$\tau\f$ is the loop delay.
        inline Complex H_sys(Real freq, Real gain)
        {
          Real arg(2*pi*freq);
          return ( gain / integrationTime ) *
            ( polar(Real(1.0), -arg*loopDelay) / Complex(0.0, arg) );
        }
        
        // Compute the transfer function of the wave front sensor, which is
        // defined as:
        // \f[
        //   H_{WFS}(f) = \mathrm{sinc}(\pi f T_e) \exp{(-i\pi f T_e)}
        // \f]
        // where \f$f\f$ is the frequency \a freq, and \f$T_e\f$ is the
        // integration time \a integrationTime.
        inline Complex H_wfs(Real freq)
        {
          Real arg = freq*integrationTime;
          return sinc(arg) * polar(Real(1.0), -pi*arg);
        }

        // Compute the open loop transfer function, which is defined as:
        // \f[
        //   H_{ol}(f,g_i) = H_{WFS}(f) H_{sys}(f,g_i)
        // \f]
        // where \f$f\f$ is the frequency \a freq, \f$g_i\f$ is the gain of
        // the electrode \a gain, \f$H_{WFS}(f)\f$ is the tranfer function of
        // the wave front sensor, and \f$H_{sys}(f,g_i)\f$ is the system
        // tranfer function.
        inline Complex H_ol(Real freq, Real gain)
        {
          return H_sys(freq,gain) * H_wfs(freq);
        }

        // Compute the noise transfer function, which is defined as:
        // \f[
        //   H_n(f,g_i) = \frac{H_{sys}(f,g_i)}{1 + H_{ol}(f,g_i)}
        // \f]
        // where \f$f\f$ is the frequency \a freq, \f$g_i\f$ is the gain of
        // the electrode \a gain, \f$H_{sys}(f,g_i)\f$ is the system transfer
        // function, and \f$H_{ol}(f,g_i)\f$ is the open loop transfer
        // function.
        inline Complex H_n(Real freq, Real gain)
        {
          return H_sys(freq, gain) / (Real(1.0) + H_ol(freq, gain));
        }

        // Compute the upper frequency that must be used in integrating the
        // noise transfer function.
        // 
        // \note The current implementation is based on estimates that were
        // taken from a graph.
        inline Real f_max(Real gain)
        {
          Assert(0 <= gain && gain <= 1.16);
          if (gain > 1.0)   return 133.6;
          if (gain > 0.5)   return 117.5;
          if (gain > 0.25)  return  42.0;
          if (gain > 0.165) return  17.0;
          if (gain > 0.116) return  10.1;
          return 7.0;
        }

      } //# namespace


      Real noiseTransfer(Real gain)
      {
        //# Step-size
        const Real df(0.05);

        //# Running sum of the total noise transfer
        Real sum(0.0);

        //# \c fmax must be calculated, based on the value of \a gain.
        Real fmax(f_max(gain));

        //# Add half a step-size to fmax so we can loop up to fmax inclusive.
        fmax += 0.5*df;

        for(Real f=df; f<fmax; f+=df) {
          sum += df * norm(H_n(f, gain));
        }

        return sum;
      }


      VectorReal noiseTransfer(const VectorReal& gain)
      {
        DBG_TRACE("Computing noise transfer function ...");

        //# Allocate a vector to hold the results
        VectorReal result(gain.shape());

        //# Loop over all elements of \a gain.
        for (Integer i = 0; i < gain.rows(); i++) {
          result(i) = noiseTransfer(gain(i));
        }

        //# Return the result.
        return result;
      }

    } //# namespace PSF

  } //# namespace SINFONI

} //# namespace ASTRON

