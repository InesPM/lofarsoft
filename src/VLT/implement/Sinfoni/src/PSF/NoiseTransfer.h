//#  NoiseTransfer.h: Global methods needed to calculate the noise transfer.
//#
//#  Copyright (C) 2002-2006
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: NoiseTransfer.h,v 1.10 2006/05/19 16:03:24 loose Exp $

#ifndef SINFONI_PSF_NOISETRANSFER_H
#define SINFONI_PSF_NOISETRANSFER_H

#include <Sinfoni/Config.h>

//# Includes
#include <Sinfoni/Types.h>

namespace ASTRON
{
  namespace SINFONI
  {
    namespace PSF
    {
      // Compute the noise transfer, which is defined as: 
      // \f[
      //   \int |H_n(g_i,f)|^2 df
      // \f] 
      // where \f$H_n(g_i,f)\f$ is the so-called noise transfer function.
      //
      // \note We took a quick and dirty approach to calculate the integral
      // \f$ \int |H_n(g_i,f)|^2 df \f$, using a fixed step-size \c df. A
      // better approach would be, of course, to use an algorithm that uses
      // user-specified boundary conditions like maximum absolute error and
      // maximum relative error (e.g. an algorithm based on the Gauss-Kronrod
      // rule).
      Real noiseTransfer(Real gain);

      // Compute the noise transfer function for a vector of gains.
      VectorReal noiseTransfer(const VectorReal& gain);

    } // namespace PSF

  } // namespace SINFONI

} // namespace ASTRON

#endif
