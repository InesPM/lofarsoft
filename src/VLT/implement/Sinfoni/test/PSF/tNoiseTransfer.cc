//#  tNoiseTransfer.cc: Check calculation of the noise transfer function.
//#
//#  Copyright (C) 2002-2006
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: tNoiseTransfer.cc,v 1.6 2006/05/19 16:03:24 loose Exp $

#include <Sinfoni/PSF/NoiseTransfer.h>
#include <Sinfoni/Exception.h>
#include <Sinfoni/Types.h>
#include <iostream>
#include <iomanip>

using namespace ASTRON::SINFONI::PSF;
using namespace ASTRON::SINFONI;
using namespace std;

int main()
{
  for(Real g=-0.01; g<1.18; g+=0.02) {
    try {
      Real Hn(noiseTransfer(g));
      cout << fixed << setprecision(2) << "noiseTransfer(" << g << ") = "
           << setw(12) << setprecision(8) << right << Hn << endl;
    } catch (Exception& e) {
      cerr << "Warning: " << e.what() << endl;
    }
  }
  return 0;
}
