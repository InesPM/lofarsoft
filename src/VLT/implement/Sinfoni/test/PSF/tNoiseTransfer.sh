#!/bin/sh
$lofar_sharedir/runtest.sh tNoiseTransfer > tNoiseTransfer_test.log 2>&1
