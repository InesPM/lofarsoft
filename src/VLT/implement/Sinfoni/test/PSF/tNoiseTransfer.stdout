noiseTransfer(0.01) =   0.81075632
noiseTransfer(0.03) =   2.24454161
noiseTransfer(0.05) =   3.37364003
noiseTransfer(0.07) =   4.23561462
noiseTransfer(0.09) =   4.88105219
noiseTransfer(0.11) =   5.36043664
noiseTransfer(0.13) =   7.14915359
noiseTransfer(0.15) =   7.65756704
noiseTransfer(0.17) =  10.74720422
noiseTransfer(0.19) =  11.49887772
noiseTransfer(0.21) =  12.16330318
noiseTransfer(0.23) =  12.74891921
noiseTransfer(0.25) =  13.26401638
noiseTransfer(0.27) =  21.24115735
noiseTransfer(0.29) =  22.56045874
noiseTransfer(0.31) =  23.83632066
noiseTransfer(0.33) =  25.06780742
noiseTransfer(0.35) =  26.25418516
noiseTransfer(0.37) =  27.39492753
noiseTransfer(0.39) =  28.48971816
noiseTransfer(0.41) =  29.53844977
noiseTransfer(0.43) =  30.54122012
noiseTransfer(0.45) =  31.49832485
noiseTransfer(0.47) =  32.41024755
noiseTransfer(0.49) =  33.27764740
noiseTransfer(0.51) =  52.47697208
noiseTransfer(0.53) =  54.89812589
noiseTransfer(0.55) =  57.35009937
noiseTransfer(0.57) =  59.83329581
noiseTransfer(0.59) =  62.34812071
noiseTransfer(0.61) =  64.89498140
noiseTransfer(0.63) =  67.47428674
noiseTransfer(0.65) =  70.08644670
noiseTransfer(0.67) =  72.73187195
noiseTransfer(0.69) =  75.41097343
noiseTransfer(0.71) =  78.12416181
noiseTransfer(0.73) =  80.87184696
noiseTransfer(0.75) =  83.65443734
noiseTransfer(0.77) =  86.47233936
noiseTransfer(0.79) =  89.32595663
noiseTransfer(0.81) =  92.21568923
noiseTransfer(0.83) =  95.14193284
noiseTransfer(0.85) =  98.10507781
noiseTransfer(0.87) = 101.10550819
noiseTransfer(0.89) = 104.14360062
noiseTransfer(0.91) = 107.21972320
noiseTransfer(0.93) = 110.33423416
noiseTransfer(0.95) = 113.48748052
noiseTransfer(0.97) = 116.67979660
noiseTransfer(0.99) = 119.91150240
noiseTransfer(1.01) = 130.83587913
noiseTransfer(1.03) = 134.62142146
noiseTransfer(1.05) = 138.47174537
noiseTransfer(1.07) = 142.38830983
noiseTransfer(1.09) = 146.37261569
noiseTransfer(1.11) = 150.42620717
noiseTransfer(1.13) = 154.55067326
noiseTransfer(1.15) = 158.74764932
